src/nb_internals.cpp

```diff
@@ -317,13 +335,19 @@ static void internals_cleanup() {
     }
 
     if (!leak) {
-        nb_translator_seq* t = internals->translators.next;
+        nb_translator_seq* t = p->translators.next;
         while (t) {
             nb_translator_seq *next = t->next;
             delete t;
             t = next;
         }
-        delete internals;
+
+#if defined(NB_FREE_THREADED)
```

Comment:
This is inside an `#if !defined(NB_FREE_THREADED)` so it's pretty much just executable documentation - might be worth a comment about that? You could add deletion of `p->shards` if you like.

Reply:
Made some tweaks here but not sure if it addresses your comment.

Reply:
This is exactly what I was imagining, thanks!